network/src/peers_registry.rs

```diff
@@ -88,259 +92,350 @@ type ProtocolConnec = (ProtocolId, UniqueConnec<(UnboundedSender<Bytes>, u8)>);
+    #[inline]
+    fn network_group(&self) -> Group {
+        self.connected_addr.network_group()
     }
 }
 
 pub struct ConnectionStatus {
     pub total: u32,
-    pub unreserved_incoming: u32,
-    pub unreserved_outgoing: u32,
-    pub max_incoming: u32,
-    pub max_outgoing: u32,
+    pub unreserved_inbound: u32,
+    pub unreserved_outbound: u32,
+    pub max_inbound: u32,
+    pub max_outbound: u32,
 }
 
 pub(crate) struct PeersRegistry {
     // store all known peers
     peer_store: Arc<RwLock<Box<PeerStore>>>,
-    peer_connections: PeerConnections,
-    // max incoming limitation
-    max_incoming: u32,
-    // max outgoing limitation
-    max_outgoing: u32,
+    peers: PeerConnections,
+    // max inbound limitation
+    max_inbound: u32,
+    // max outbound limitation
+    max_outbound: u32,
     // Only reserved peers or allow all peers.
     reserved_only: bool,
-    deny_list: PeersDenyList,
+    reserved_peers: FnvHashSet<PeerId>,
+}
+
+fn find_most_peers_in_same_network_group<'a>(
+    peers: impl Iterator<Item = (&'a PeerId, &'a PeerConnection)>,
+) -> Vec<&'a PeerId> {
+    peers
+        .fold(
+            FnvHashMap::with_capacity_and_hasher(16, Default::default()),
+            |mut groups, (peer_id, peer)| {
+                groups
+                    .entry(peer.network_group())
+                    .or_insert_with(Vec::new)
+                    .push(peer_id);
+                groups
+            },
+        )
+        .values()
+        .max_by_key(|group| group.len())
+        .cloned()
+        .unwrap_or_else(Vec::new)
+}
+
+fn sort_then_drop_last_n_elements<T, F>(list: &mut Vec<T>, n: usize, compare: F)
+where
+    F: FnMut(&T, &T) -> std::cmp::Ordering,
+{
+    list.sort_by(compare);
+    list.truncate(list.len().saturating_sub(n));
 }
 
 impl PeersRegistry {
     pub fn new(
         peer_store: Arc<RwLock<Box<PeerStore>>>,
-        max_incoming: u32,
-        max_outgoing: u32,
+        max_inbound: u32,
+        max_outbound: u32,
         reserved_only: bool,
+        reserved_peers: Vec<PeerId>,
     ) -> Self {
-        let deny_list = PeersDenyList::new();
+        let mut reserved_peers_set =
+            FnvHashSet::with_capacity_and_hasher(reserved_peers.len(), Default::default());
+        for reserved_peer in reserved_peers {
+            reserved_peers_set.insert(reserved_peer);
+        }
         PeersRegistry {
             peer_store,
-            peer_connections: Default::default(),
-            max_incoming,
-            max_outgoing,
+            peers: Default::default(),
+            reserved_peers: reserved_peers_set,
+            max_inbound,
+            max_outbound,
             reserved_only,
-            deny_list,
         }
     }
 
     #[allow(clippy::needless_lifetimes)]
     #[inline]
     pub fn get_peer_id<'a>(&'a self, peer_index: PeerIndex) -> Option<&'a PeerId> {
-        self.peer_connections.get_peer_id(peer_index)
+        self.peers.get_peer_id(peer_index)
```

Comment:
```
    #[inline]
    fn get_peer_id(&self, peer_index: PeerIndex) -> Option<&PeerId> {
        self.peer_id_by_index.get(&peer_index)
    }

    #[inline]
    fn get_mut(&mut self, peer_id: &PeerId) -> Option<&mut PeerConnection> {
        self.peers.get_mut(peer_id)
    }

    #[inline]
    fn iter(&self) -> impl Iterator<Item = (&PeerId, &PeerConnection)> {
        self.peers.iter()
    }
```
any problem with clippy lint here？ why  dismiss it.

Reply:
Clippy will output a false warning in the previous version, seem fixed in this version